R/derive_merged.R

```diff
@@ -0,0 +1,708 @@
+#'
+#' @param dataset_add Additional dataset
+#'
+#'   The variables specified by the `by_vars`, the `dtc`, and the `order`
+#'   parameter are expected.
+#'
+#' @inheritParams derive_vars_merged
+#' @inheritParams derive_vars_dt
+#'
+#' @details
+#'
+#'   1. The additional dataset is restricted to the observations matching the
+#'   `filter_add` condition.
+#'
+#'   1. The date variable and if requested, the date imputation flag is added to
+#'   the additional dataset.
+#'
+#'   1. If `order` is specified, for each by group the first or last observation
+#'   (depending on `mode`) is selected.
+#'
+#'   1. The date and flag variables are merged to the input dataset.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @keywords derivation adam timing
+#'
+#' @export
+#'
+#' @examples
+#' library(admiral.test)
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("dm")
+#' data("ex")
+#'
+#' # derive treatment start date (TRTSDT)
+#' derive_vars_merged_dt(
+#'   select(dm, STUDYID, USUBJID),
+#'   dataset_add = ex,
+#'   by_vars = vars(STUDYID, USUBJID),
+#'   new_vars_prefix = "TRTS",
+#'   dtc = EXSTDTC,
+#'   date_imputation = "first",
+#'   order = vars(TRTSDT),
+#'   mode = "first"
+#' )
+#'
+#' # derive treatment end date (TRTEDT) (without imputation)
+#' derive_vars_merged_dt(
+#'   select(dm, STUDYID, USUBJID),
+#'   dataset_add = ex,
+#'   by_vars = vars(STUDYID, USUBJID),
+#'   new_vars_prefix = "TRTE",
+#'   dtc = EXENDTC,
+#'   order = vars(TRTEDT),
+#'   mode = "last"
+#' )
+derive_vars_merged_dt <- function(dataset,
+                                  dataset_add,
+                                  by_vars,
+                                  order = NULL,
+                                  new_vars_prefix,
+                                  filter_add = NULL,
+                                  mode = NULL,
+                                  dtc,
+                                  date_imputation = NULL,
+                                  flag_imputation = TRUE,
+                                  min_dates = NULL,
+                                  max_dates = NULL,
+                                  check_type = "warning",
+                                  duplicate_msg = NULL) {
+  dtc <- assert_symbol(enquo(dtc))
+  filter_add <- assert_filter_cond(enquo(filter_add), optional = TRUE)
+  assert_data_frame(dataset_add, required_vars = quo_c(by_vars, dtc))
+
+  old_vars <- names(dataset_add)
+  add_data <- filter_if(dataset_add, filter_add) %>%
+    derive_vars_dt(
+      new_vars_prefix = new_vars_prefix,
+      dtc = !!dtc,
+      date_imputation = date_imputation,
+      flag_imputation = flag_imputation,
+      min_dates = min_dates,
+      max_dates = max_dates
+    )
+  new_vars <- quos(!!!syms(setdiff(names(add_data), old_vars)))
+  derive_vars_merged(dataset,
+                     dataset_add = add_data,
+                     by_vars = by_vars,
+                     order = order,
```

Comment:
When `date_imputation` is NULL and a partial date exists, then the `derive_vars_dt` step converts the partial dates to missing. As a result, the date variable specified in `order` dictates which record is chosen. For example, if there are two records with `EXSTDTC` being`“2020”` and `“2020-12-14”`, then converted `TRTSDT` are `NA` and `2020-12-14`. Assume `mode = “first”`, if `order = vars(TRTSDT)` (converted date) then the record with complete date is chosen. But if `order = vars(EXSTDTC)` (unconverted character date), then the record with partial date is chosen. I’m not sure if this is intended or I'm making this too confusing. If it's intended, should we include some clarifications about which records are chosen when partial dates are not imputed?

Reply:
Good point! I have added some details to the description of the `order` parameter. I would expect that in most cases missings are excluded by `filter_add`.

However, the current behaviour looks a bit inconsistent. For `mode = "first"` observations with `NA` are selected only if all observation are `NA` while for `mode = "last"` observations with `NA` are always selected. Maybe we should change this behaviour. What do you think?

Reply:
Thanks for adding the details in the description. 

If I'm understanding correctly, this behavior seems to be carried from `filter_extreme`. I think in some cases the users may not want to choose records with `NA`. Maybe we can add a parameter for `filter_extreme` (e.g. `remove_na`) to allow users to choose whether `NA` should be considered?

Reply:
Yes, you are correct. I have added it to the agenda of the next "Questions and Comments" meeting as it affects many functions.

Reply:
Hi @yangc71 , we discussed the topic today and decided to keep the behavior as it is because it is in line with the `arrange()` function. We also discussed to provide a `remove_na` parameter but decided not to provide it because usually `filter_extreme` is not called directly but within a derivation and the derivations already provide a `filter` parameter to exclude NAs.